Shopify/tophat
# Restructure download and install mechanisms to cache artifacts

### What does this change accomplish?

This change modifies the mechanisms used for downloading and installing applications so that artifacts are cached within a 30 second rolling window.  This makes retries or repeated reinstalls much faster.

### How have you achieved it?

By creating some entities:
- `CachingApplicationDownloader` accepts remote sources and returns applications for them.  Internally, this makes use of `ArtifactDownloader` and `ArtifactUnpacker`.
- `ArtifactContainer` represents a filesystem node that (currently) can contain a raw download and an application.  Download and unpack operations take place within this container.  The container is destroyed when it is no longer needed.
- `InstallSession` represents a session within the rolling window where downloads and installations take place.  When an instance of this actor is destroyed, all of its temporary resources are also destroyed.

`InstallCoordinator` has been modified to track the lifecycle of `InstallSession`s and delegates to the current session as needed.

### How can the change be tested?

1. Select two iOS simulators simultaneously.
2. Install an iOS simulator build.  Observe that it is only downloaded once, then the same build is installed to both simulators.
3. Quickly install the same build again.  It should be very quick.
4. Wait 30 seconds and try again.  The artifact should be downloaded again like in step 2.
